### PR TITLE
add support for docker images and PIM instances in libuserd

### PR DESCRIPTION
**Pull request by mariostieriansys**
The libuserd was missing support for docker images and/or PIM instances (like on Ansys Lab).

This PR adds this feature. What was required:

1) Make some of the stuff used in DockerLauncher common, hence I introduced a common module with stuff that can be used around. I believe we can keep using it for other methods in the future
2) Add (unfortunately) a lot of new parameters in the __init__ to change the behavior depending on the image, and if PIM is involved or not
3) Add methods to connect to enshell and launch the container if required
4) Finally, add a wrapper in launch_ensight, called launch_libuserd, which simplifies the launch when PIM is involved, like we have for EnSight on Ansys Lab

I also introduced a download_pyansys_example interface which allows the users to get stuff from the ansys/example-data repository. We can so use it to get data at runtime when running github tests

### CODE REVIEW

File: tests/unit_tests/test_dockerlauncher.py [hunk 1]
```diff
@@ -90,15 +91,6 @@ def test_start(mocker, capsys, caplog, enshell_mock, tmpdir):
     values_run_command[0] = [0, "set_no_reroute_log"]
     enshell_mock[0].run_command.side_effect = values_run_command.copy()
     mocker.patch.object(enshell_grpc, "EnShellGRPC", return_value=enshell_mock[0])
-    dock = mocker.patch.object(docker, "from_env", return_value=docker_client)
```
Comment (mariostieriansys) [inline tests/unit_tests/test_dockerlauncher.py:93]: In the PR I removed the check for the docker module.
We are always installing it, there's no need to check if the module loaded or not. 
The docker module is then used only if DockerLauncher is started, or docker is asked for in LibUserd. If docker is not installed on the machine where these are being asked for, it is correct that the docker-py module should fail miserably.
So these checks are no more required

File: CHANGELOG.rst [hunk 2]
```diff
@@ -1,6 +1,6 @@
-=======
```
Comment (mariostieriansys) [inline CHANGELOG.rst:1]: the new pre-commit check for merge conflict doesn't like the =======.... it assumes there's a conflict there

**Review by randallfrank — APPROVED** [review]

File: CHANGELOG.rst (hunk 2 above)
Comment (randallfrank) [reply CHANGELOG.rst:1]: Check the sphinx output on this one.  Sphinx is a little picky about the number of chars matching the number of chars in the string the formatting is being applied to.   It is not an error to sphinx, but it usually generates a warning.

Comment (mariostieriansys) [reply CHANGELOG.rst:1]: I have checked the rendering of the Queries example and it works well.
The changelog is generated only on the repo page, so I cannot see it directly. However, I have tried different RST viewers and it gets rendered correctly, so I believe we should be fine

Comment (mariostieriansys) [reply CHANGELOG.rst:1]: image

**Pull request merged**